src/Plutarch/Test/QuickCheck.hs

```diff
@@ -77,120 +84,226 @@ import Test.QuickCheck (
 
--- | @since 2.0.0
-instance
-    forall
-        (a :: S -> Type)
-        (b :: S -> Type)
-        (c :: S -> Type)
-        (d :: S -> Type)
-        (e :: S -> Type).
-    ( c ~ (d :--> e)
-    , PLift a
-    , PLift b
-    , FromPFunN d e (PTT c)
-    ) =>
-    FromPFunN (a :--> b) c 'PFunction
-    where
-    fromPFun' _ f (PFun _ _ (unTestableTerm -> x)) =
-        fromPFun' (Proxy @(PTT c)) $ f # x
+{- | Constraint for `PUnLam'` that will give a better type error message.
 
--- | @since 2.0.0
-instance
-    forall (a :: S -> Type) (b :: S -> Type) (c :: S -> Type) (d :: S -> Type).
-    ( b ~ (c :--> d)
-    , FromPFunN c d (PTT b)
-    ) =>
-    FromPFunN a b 'PTerm
-    where
-    fromPFun' _ f (TestableTerm x) = fromPFun' (Proxy @(PTT b)) $ f # x
+ @since 2.1.0
+-}
+type PUnLam fin p = (PUnLam' fin p (IsFinal fin p), CheckReturn fin p)
+
+{- | Wrapper for @pUnLam'@. Only required the final return type.
+
+ @since 2.1.0
+-}
+punlam' ::
+    forall (fin :: S -> Type) (p :: S -> Type).
+    PUnLam fin p =>
+    (forall s. Term s p) ->
+    PUnLamHask fin p
+punlam' = pUnLam' @fin @p @(IsFinal fin p)
+
+{- | Wrapper for @pUnLam'@. Same as @punlam'@ but evaluates the given
+     Plutarch function before the conversion. It will throw an error
+     if evaluation failes.
+
+ @since 2.1.0
+-}
+punlam ::
+    forall (fin :: S -> Type) (p :: S -> Type).
+    PUnLam fin p =>
+    (forall s. Term s p) ->
+    PUnLamHask fin p
+punlam pf = punlam' @fin (loudEval pf)
 
-{- | Converts Plutarch Functions into `Testable` Haskell function of
-  TestableTerms.
+{- | Converts Plutarch Functions into Haskell function of
+     TestableTerms. Then, wrap any instance of TestableTerm that is
+     lambda into `PFun`. Result of this function will be
+     "QuickCheck-compatible" given that all Plutarch type used have
+     `PArbitrary` instance.
 
  @since 2.0.0
 -}
 fromPFun ::
-    forall (a :: S -> Type) (b :: S -> Type) (c :: PTermType).
-    ( c ~ PTT (a :--> b)
-    , FromPFunN a b c
+    forall (p :: S -> Type).
+    ( PUnLam PBool p
+    , PWrapLam (PUnLamHask PBool p)
     ) =>
-    (forall s. Term s (a :--> b)) ->
-    TestableFun c (a :--> b)
-fromPFun = fromPFun' (Proxy @(PTT (a :--> b)))
+    ClosedTerm p ->
+    PLamWrapped (PUnLamHask PBool p)
+fromPFun pf = pwrapLam $ punlam @PBool pf
+
+{- | Ways an Plutarch terms can be compared.
+     @OnPEq@ uses Plutarch `PEq` instance to compare give terms. This
+     means two terms with different UPLC representations can be
+     considered equal when `PEq` instance defines so.
+     @OnUPLC@ uses compiled and evaluated raw UPLC to compare two
+     terms. It is useful comparing Terms that forgot their types--
+     `POpqaue`.
+
+ @since 2.1.0
+-}
+data Equality
+    = OnPEq
+    | OnPData
+    | OnBoth
+    deriving stock (Eq, Show)
+
+{- | Partiality of the comparison. @ByPartial@ will have some
+     performance disadventages.
+
+ @since 2.1.0
+-}
+data Partiality
+    = ByComplete
+    | ByPartial
```

Comment:
Again, what are the semantics here? From what I can see, we have the following combinations:

* `OnPEq` + `ByComplete`
* `OnPEq` + `ByPartial`
* `OnPData` + `ByComplete`
* `OnPData` + `ByPartial`
* `OnBoth` + `ByComplete`
* `OnBoth` + `ByPartial`

What's the difference between all these?

Reply:
`OnPEq` compares with `PEq` instance, and `OnPData` only compares the `PData`--It requires `PIsData` instance. `OnBoth` checks `OnPEq` *and* `OnPData`. 

`ByPartial` allows user to provide a partial Haskell function that returns `Maybe`, while `OnComplete` does not. This is separated because `ByPartial` is slower and We don't want to require every function tested to be partial and therefore slower.

Reply:
How does `OnBoth` check both? I assume it's a conjunction: meaning that they must both be `True` to succeed? I guess this also translates to the partial case, in which case they must both be `Just True`?

Reply:
yes `OnBoth` conjoins:
```hs
instance
    forall (p :: S -> Type) (h :: Type).
    (PLamArgs p ~ '[], PLift p, PLifted p ~ h, PIsData p, PEq p) =>
    HaskEquiv 'OnBoth 'ByComplete h p '[]
    where
    haskEquiv h p _ =
        (haskEquiv @( 'OnPEq) @( 'ByComplete) h p Nil) .&&. (haskEquiv @( 'OnPData) @( 'ByComplete) h p Nil)
```

When it is combined with `ByPartial` it just means: "When Haskell function returns Nothing, given plutarch function should not evaluate--and it will only run it once. When Haskell function gives Just x, Plutarch function should return a value that matches "x" by PEq and by PData comparison

Reply:
I think those semantics are fine, but they should _definitely_ be documented more clearly.

Reply:
Okay. Sounds good. I kind of assumed that made sense :D

Reply:
It does make sense, but only upon reading the code, carefully. The Haddocks should speak for themselves in this regard for users, as they shouldn't have to read our code to understand what it's doing.